### PR TITLE
[stdlib] Make various free functions mutating an _ArrayBufferProtocol into extensions

### PR DESCRIPTION
**Pull request by airspeedswift**
This started from the fairly innocent `FIXME(ABI)#14 : add argument labels to conform to naming convention guidelines.` but then evolved into changing a number of generic functions that took an `inout` conforming to `_ArrayBufferProtocol` to be mutating methods on the protocol instead.

Also added a missing `endIndex: Int` declaration to the protocol, which was what appears to have been forcing the need to keep adding `where Index == Int` in various places, and did some other cleanup to make `replace(subRange:)` look more like `RangeReplaceableCollection`'s `replaceSubrange`.

This seems to leave us within striking distance of making `_ArrayBufferProtocol` conform to `RangeReplaceableCollection`, which would tidy up a number of functions that are duplicating functionality that exists there. But that may have performance impact so putting this in first.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please test

File: stdlib/public/core/Arrays.swift.gyb [hunk 1]
```diff
@@ -1823,141 +1819,141 @@ public func += <
 
 //===--- generic helpers --------------------------------------------------===//
```
Comment (moiseev) [inline stdlib/public/core/Arrays.swift.gyb]: They are not really generic anymore... in a sense.

**Comment by moiseev** [comment]
@swift-ci Please benchmark

**Comment by airspeedswift** [comment]
Benchmarks look ok to me.

**Comment by slavapestov** [comment]
Cool!

**Pull request merged**